libmamba/src/core/curl.cpp

```diff
@@ -307,4 +308,180 @@ namespace mamba
         return m_errorbuffer;
     }
 
+    CURL* unwrap(const CURLHandle& h)
+    {
+        return h.m_handle;
+    }
+
+    bool operator==(const CURLHandle& lhs, const CURLHandle& rhs)
+    {
+        return unwrap(lhs) == unwrap(rhs);
+    }
+
+    bool operator!=(const CURLHandle& lhs, const CURLHandle& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    /*****************
+     * CURLReference *
+     *****************/
+
+    CURLReference::CURLReference(CURL* handle)
+        : p_handle(handle)
+    {
+    }
+
+    CURL* unwrap(const CURLReference& h)
+    {
+        return h.p_handle;
+    }
+
+    bool operator==(const CURLReference& lhs, const CURLReference& rhs)
+    {
+        return unwrap(lhs) == unwrap(rhs);
+    }
+
+    bool operator==(const CURLReference& lhs, const CURLHandle& rhs)
+    {
+        return unwrap(lhs) == unwrap(rhs);
+    }
+
+    bool operator==(const CURLHandle& lhs, const CURLReference& rhs)
+    {
+        return unwrap(lhs) == unwrap(rhs);
+    }
+
+    bool operator!=(const CURLReference& lhs, const CURLReference& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    bool operator!=(const CURLReference& lhs, const CURLHandle& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    bool operator!=(const CURLHandle& lhs, const CURLReference& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    /*******************
+     * CURLMultiHandle *
+     *******************/
+
+    CURLMultiHandle::CURLMultiHandle(std::size_t max_parallel_downloads)
+        : p_handle(curl_multi_init())
+        , m_max_parallel_downloads(max_parallel_downloads)
+    {
+        if (p_handle == nullptr)
+        {
+            throw curl_error("Could not initialize CURL multi handle");
+        }
+        else
+        {
+            curl_multi_setopt(
+                p_handle,
+                CURLMOPT_MAX_TOTAL_CONNECTIONS,
+                static_cast<int>(max_parallel_downloads)
+            );
+        }
+    }
+
+    CURLMultiHandle::~CURLMultiHandle()
+    {
+        curl_multi_cleanup(p_handle);
+        p_handle = nullptr;
+    }
+
+
+    CURLMultiHandle::CURLMultiHandle(CURLMultiHandle&& rhs)
+        : p_handle(rhs.p_handle)
+        , m_max_parallel_downloads(rhs.m_max_parallel_downloads)
+    {
+        rhs.p_handle = nullptr;
+        rhs.m_max_parallel_downloads = 0u;
+    }
+
+    CURLMultiHandle& CURLMultiHandle::operator=(CURLMultiHandle&& rhs)
+    {
+        std::swap(p_handle, rhs.p_handle);
```

Comment:
Shouldnt we set `rhs` in a non-owning state? Just swapping here will make it own what we have, it might not be immediately destroyed.

Reply:
It's ok to have rhs freeing the curl handle in its destructor later.

Reply:
Yes, the object it is swapped with is in a valid state as long as it has been constructed.